lib/HLSL/HLModule.cpp

```diff
@@ -1028,51 +1033,52 @@ static Value *MergeGEP(GEPOperator *SrcGEP, GetElementPtrInst *GEP) {
     Indices.append(SrcGEP->op_begin() + 1, SrcGEP->op_end());
     Indices.append(GEP->idx_begin() + 1, GEP->idx_end());
   }
-  if (!Indices.empty())
-    return Builder.CreateInBoundsGEP(SrcGEP->getSourceElementType(),
-                                     SrcGEP->getOperand(0), Indices,
-                                     GEP->getName());
-  else
-    llvm_unreachable("must merge");
+
+  DXASSERT(!Indices.empty(), "must merge");
+  Value *newGEP = Builder.CreateInBoundsGEP(nullptr, SrcGEP->getOperand(0), Indices, Name);
+  GEP->replaceAllUsesWith(newGEP);
+  if (Instruction *I = dyn_cast<Instruction>(GEP))
+    I->eraseFromParent();
+  return newGEP;
 }
 
 void HLModule::MergeGepUse(Value *V) {
-  for (auto U = V->user_begin(); U != V->user_end();) {
-    auto Use = U++;
-
-    if (GetElementPtrInst *GEP = dyn_cast<GetElementPtrInst>(*Use)) {
-      if (GEPOperator *prevGEP = dyn_cast<GEPOperator>(V)) {
-        // merge the 2 GEPs
-        Value *newGEP = MergeGEP(prevGEP, GEP);
-        // Don't need to replace when GEP is updated in place
-        if (newGEP != GEP) {
-          GEP->replaceAllUsesWith(newGEP);
-          GEP->eraseFromParent();
-        }
-        MergeGepUse(newGEP);
+  SmallVector<Value*, 16> worklist;
```

Comment:
New worklist pattern is safer since we will now be potentially modifying and adding users in ways the recursive iteration pattern could not handle.  It's also a lot easier to follow this way.